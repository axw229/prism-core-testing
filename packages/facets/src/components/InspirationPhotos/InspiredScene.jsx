// @flow
import React, { useContext } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import Prism, { ColorPin, ImageColorPicker } from '@prism/toolkit'
import ConfigurationContext, { type ConfigurationContextType } from 'src/contexts/ConfigurationContext/ConfigurationContext'
import { fullColorNumber, getContrastYIQ } from 'src/shared/helpers/ColorUtils'
import { add } from 'src/store/actions/live-palette'
import useColors from '../../shared/hooks/useColors'

type InspiredSceneProps = {
  data: { img: string, initPins: { x: number, y: number }[] }
}

const InspiredScene = ({ data: { img, initPins } }: InspiredSceneProps) => {
  // We call useColor to load the colors for cases where loadColor has not already been called.
  // we don't use the colors from what is returned bc the data type for unorderedColors (sting[]) is much thinner in the hook
  // than what is in redux (Color[])
  // We consume the colors from the redux store, since it's a bit redundant to recreate what we already have access to
  // eslint-disable-next-line no-unused-vars
  const [colors] = useColors()
  const unorderedColors = useSelector(store => store.colors?.unorderedColors)
  const dispatch = useDispatch()
  const livePaletteColors = useSelector((store) => store.lp.colors)
  const { brandKeyNumberSeparator, colorWall: { colorSwatch = {} } } = useContext<ConfigurationContextType>(ConfigurationContext)
  const { colorNumOnBottom = false, houseShaped = false } = colorSwatch
  const baseClass = `image-color-picker${colorNumOnBottom ? `--name-number${houseShaped ? '-house-shaped' : ''}` : '--number-name'}`

  const isColorAdded = ({ colorNumber }): boolean => livePaletteColors.some((c) => c.colorNumber === colorNumber)

  return (
    <Prism style={{ padding: '0 0.125rem' }}>
      {unorderedColors && (
        <ImageColorPicker
          colors={unorderedColors}
          imgSrc={img}
          initialPinLocations={initPins.map((pin) => ({ x: pin.x * 100, y: pin.y * 100 }))}
          pinRenderer={(props) => {
            return <ColorPin
              {...props}
              buttonContent={(color) => (
                <FontAwesomeIcon
                  aria-label={isColorAdded(color) ? 'added' : 'add'}
                  icon={isColorAdded(color) ? ['fa', 'check-circle'] : ['fal', 'plus-circle']}
                  style={{ color: getContrastYIQ(color.hex) }}
                />
              )}
              labelContent={(color) => (
                <div className={baseClass}>
                  <p className={`${baseClass}__number`}>{fullColorNumber(color.brandKey, color.colorNumber, brandKeyNumberSeparator)}</p>
                  <p className={`${baseClass}__name`}>{color.name}</p>
                </div>
              )}
              onColorAdded={(color) => dispatch(add(color))}
            />
          }}
          removeButtonContent={
            <FontAwesomeIcon aria-label='remove' icon={['fa', 'trash']} style={{ display: 'inline-block' }} />
          }
        />
      )}
    </Prism>
  )
}

export default InspiredScene
