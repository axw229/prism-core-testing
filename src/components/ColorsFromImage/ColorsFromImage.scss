.match-photo-picker {
  &__wrapper {
    width: 100%;
    /* @todo - This just stopped the annoy jank, needs to be properly fixed. -RS */
    cursor: pointer;
  }
  &__canvas {
    &--portrait {
      margin: 0 auto;
      text-align: center;
      display: block;
    }
  }
}