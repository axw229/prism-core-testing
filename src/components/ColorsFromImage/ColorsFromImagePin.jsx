// @flow
import React, { PureComponent, Fragment } from 'react'

type Props = {
  pinType: string,
  onClickMethod: Function,
  // activePinIndex: number,
  isActiveFlag: boolean,
  transformValue: string,
  RGBstring: string,
  previewColorName: string,
  previewColorNumber: string
}

class ColorsFromImagePin extends PureComponent<Props> {
  render () {
    // eslint-disable-next-line one-var
    const { previewColorName, previewColorNumber, transformValue, RGBstring, onClickMethod, isActiveFlag, pinType } = this.props

    let isPreviewPin = false
    let additionalPinClass = ''
    let isActiveClass = ''

    if (pinType === 'preview') {
      isPreviewPin = true
      additionalPinClass = 'pin--preview__wrapper'
    }

    if (isPreviewPin) {
      if (isActiveFlag) isActiveClass = 'pin--active__wrapper'
    } else {
      // TODO: fix this functionality
      // is `this.key` supposed to be a reference to the key prop? is it supposed to be internally tracked?
      // if (activePinIndex === this.key) isActiveClass = 'pin--active__wrapper'
    }

    return (
      <Fragment>
        <button onClick={onClickMethod} style={{ transform: transformValue }} className={additionalPinClass + ' pin__wrapper ' + isActiveClass}>
          <span className='pin__chip' style={{ background: RGBstring }} />
          <div className='pin__name-wrapper'>
            <span className='pin__copy pin__name'>{previewColorName}</span>
            <span className='pin__copy pin__number'>SW {previewColorNumber}</span>
          </div>
        </button>
      </Fragment>
    )
  }
}

export default ColorsFromImagePin
