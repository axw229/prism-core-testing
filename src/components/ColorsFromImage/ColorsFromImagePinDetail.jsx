<div class="color-detail__wrapper">
  <div class="color-detail">
    <div class="scene__room__tinted-wrapper">
      <img class="scene__room scene__room--tinted" src="./images/room.jpg" alt="" />
    </div>
    <img class="scene__room scene__room--natural" src="./images/room.jpg" alt="" />
  </div>
</div>