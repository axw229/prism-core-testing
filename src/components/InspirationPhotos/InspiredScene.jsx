// @flow
import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { add } from 'src/store/actions/live-palette'
import Prism, { ColorPin, ImageColorPicker } from 'prism-lib'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faPlusCircle } from '@fortawesome/pro-light-svg-icons'
import { faCheckCircle, faTrash } from '@fortawesome/pro-solid-svg-icons'
import { getContrastYIQ } from '../../../src/shared/helpers/ColorUtils'

type InspiredSceneProps = {
  data: { img: string, initPins: { x: number, y: number }[] }
}

const InspiredScene = ({ data: { img, initPins } }: InspiredSceneProps) => {
  const colors = useSelector((store) => store.colors.unorderedColors)
  const dispatch = useDispatch()
  const livePaletteColors = useSelector((store) => store.lp.colors)

  const isColorAdded = ({ colorNumber }): boolean => livePaletteColors.some((c) => c.colorNumber === colorNumber)

  return (
    <Prism style={{ padding: '0 0.125rem' }}>
      {colors && (
        <ImageColorPicker
          colors={colors}
          imgSrc={img}
          initialPinLocations={initPins.map((pin) => ({ x: pin.x * 100, y: pin.y * 100 }))}
          pinRenderer={(props) => (
            <ColorPin
              {...props}
              buttonContent={(color) => (
                <FontAwesomeIcon
                  aria-label={isColorAdded(color) ? 'added' : 'add'}
                  icon={isColorAdded(color) ? faCheckCircle : faPlusCircle}
                  style={{ color: getContrastYIQ(color.hex) }}
                />
              )}
              onColorAdded={(color) => dispatch(add(color))}
            />
          )}
          removeButtonContent={
            <FontAwesomeIcon aria-label='remove' icon={faTrash} style={{ display: 'inline-block' }} />
          }
        />
      )}
    </Prism>
  )
}

export default InspiredScene
