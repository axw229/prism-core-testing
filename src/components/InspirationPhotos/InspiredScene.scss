.scene__image__wrapper{
    min-height: 100%;
    outline: none;
    position: relative;
    width: 100%;
    &__canvas{
        border-left:2px solid #fff;
        border-right:2px solid #fff;
        min-height: 266px;
        width:100%;
        @include respond-above(xs) {
            min-height:366px;
        }
        @include respond-above(md){
            min-height:540px;
        }
        &--portrait {
            width: auto;
            margin: 0 auto;
            display: block;
            border: none;
        }
    }
    &__image{
        height:0;
        width:100%;
        &--hidden {
            visibility: hidden;
            display: none;
        }
    }
    &__indicator{
        border:2px solid #ffffff;
        border-radius: 2em;
        cursor: pointer;
        height: 3em;
        position: absolute;
        width:3em;
        z-index: 1000;
    }
    &__delete-pin {
        background: white;
        border: 1px solid #f2f2f2;
        border-radius: 50%;
        bottom: 8%;
        color: black;
        cursor: default;
        display: flex;
        fill: black;
        height: 3em;
        left: 50%;
        outline: 0;
        padding: 1em;
        position: absolute;
        text-align: center;
        text-transform: uppercase;
        width: 3em;
        &--active{
            transform: scale(1.5);
            transition: all 0.3s;
            background-color: #2cabe2;
            border-color: #2cabe2;
            color:#ffffff;
        }
        &--hide{
            display: none;
        }
        &:focus {
            outline-color: #2cabe2;
            outline-style: solid;
            outline-width: 2px;
        }
        &:hover {
            border-color: #2cabe2;
            color: #2cabe2;
            cursor: pointer;
            fill: #2cabe2;
        }
    }
}

.colorfrom__image__wrapper {
    margin: 0;
    width: 100%;
    &__canvas{
        min-height: 266px;
        width:auto;
        display: block;
        margin: 0 auto;
        @include respond-above(xs) {
            min-height:366px;
        }
        @include respond-above(md){
            min-height:540px;
        }
    }
    &__image {
        height: 0;
        width: 100%;

        &--hidden {
            visibility: hidden;
            display: none;
        }
    }
}
