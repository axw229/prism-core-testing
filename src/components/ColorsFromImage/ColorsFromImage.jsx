'use strict'

import React, { PureComponent, Fragment, createRef } from 'react'
import { brandColors } from './sw-colors-in-LAB.js'
import { getDeltaE00 } from 'delta-e'
import ColorsFromImagePin from './ColorsFromImagePin'

import './ColorsFromImage.scss'

const baseClass = 'match-photo-picker'
const wrapperClass = `${baseClass}__wrapper`
const canvasClass = `${baseClass}__canvas`
const hiddenImageClass = `${baseClass}__hidden-image`
const portraitOrientation = `${canvasClass}--portrait`

class ColorsFromImage extends PureComponent {
  constructor (props) {
    super(props)

    this.state = {
      previewPinIsUpdating: false,
      previewPinIsActive: false,
      previewColorName: '',
      previewColorNumber: '',
      cursorIsPaused: false,
      cursorX: 0,
      cursorY: 0,
      previewPinX: 0,
      previewPinY: 0,
      mappedCanvasIndex: 0,
      currentPixelRGB: [0, 0, 0],
      currentPixelRGBstring: 'rgb(0,0,0)',
      currentBrandColorIndex: null,
      pinnedColors: [],
      activePinIndex: -1,
      currentPinIndex: -1,
      imageStatus: 'loading'
    }

    this.findBrandColorTimeout = null
    this.brandColorsLength = brandColors.length
    this.CFICanvas = createRef()
    this.CFIWrapper = createRef()
    this.imageRef = createRef()
    this.handleImageLoaded = this.handleImageLoaded.bind(this)
    this.handleImageErrored = this.handleImageErrored.bind(this)
    this.initCanvas = this.initCanvas.bind(this)
    this.wrapperMouseMove = this.wrapperMouseMove.bind(this)
    this.rgb2lab = this.rgb2lab.bind(this)
    this.findBrandColor = this.findBrandColor.bind(this)
    this.updatePreviewPin = this.updatePreviewPin.bind(this)
  }

  rgb2lab (rgb) {
    // eslint-disable-next-line one-var
    var r = rgb[0] / 255,
      g = rgb[1] / 255,
      b = rgb[2] / 255,
      x, y, z

    r = (r > 0.04045) ? Math.pow((r + 0.055) / 1.055, 2.4) : r / 12.92
    g = (g > 0.04045) ? Math.pow((g + 0.055) / 1.055, 2.4) : g / 12.92
    b = (b > 0.04045) ? Math.pow((b + 0.055) / 1.055, 2.4) : b / 12.92

    x = (r * 0.4124 + g * 0.3576 + b * 0.1805) / 0.95047
    y = (r * 0.2126 + g * 0.7152 + b * 0.0722) / 1.00000
    z = (r * 0.0193 + g * 0.1192 + b * 0.9505) / 1.08883

    x = (x > 0.008856) ? Math.pow(x, 1 / 3) : (7.787 * x) + 16 / 116
    y = (y > 0.008856) ? Math.pow(y, 1 / 3) : (7.787 * y) + 16 / 116
    z = (z > 0.008856) ? Math.pow(z, 1 / 3) : (7.787 * z) + 16 / 116

    return [(116 * y) - 16, 500 * (x - y), 200 * (y - z)]
  }

  handleImageLoaded () {
    this.setState({ imageStatus: 'loaded' })
    this.initCanvas()
  }

  handleImageErrored () {
    this.setState({ imageStatus: 'failed' })
  }

  initCanvas () {
    this.CFICanvasContext = this.CFICanvas.current.getContext('2d')
    // eslint-disable-next-line react/prop-types
    this.handleResize(this.props.width, this.props.height)
  }

  addPin () {
    console.log('Adding pin')
  }

  handleResize (width, height) {
    // @todo - determine if I need to set/get the pixeldata
    this.CFICanvasContext.clearRect(0, 0, width, height)
    // eslint-disable-next-line react/prop-types
    this.CFICanvasContext.drawImage(this.imageRef.current, 0, 0, width, height)

    this.canvasOffset = this.CFICanvas.current.getBoundingClientRect()
    this.wrapperOffset = this.CFIWrapper.current.getBoundingClientRect()
  }

  findBrandColor () {
    let currentPixelInLABarray = this.rgb2lab(this.state.currentPixelRGB)
    let currentPixelInLAB = { L: currentPixelInLABarray[0], A: currentPixelInLABarray[1], B: currentPixelInLABarray[2] }

    for (var arrayIndex = 0; arrayIndex < this.brandColorsLength; arrayIndex += 6) {
      let thisSWinLAB = { L: brandColors[arrayIndex + 3], A: brandColors[arrayIndex + 4], B: brandColors[arrayIndex + 5] }
      let colorDistance = getDeltaE00(currentPixelInLAB, thisSWinLAB)

      if (colorDistance < 12) {
        this.setState({ currentBrandColorIndex: arrayIndex, currentPixelRGB: brandColors[arrayIndex + 2], currentPixelRGBstring: 'rgb(' + brandColors[arrayIndex + 2] + ')', previewPinIsActive: true, previewColorName: brandColors[arrayIndex], previewColorNumber: brandColors[arrayIndex + 1] })
        arrayIndex = this.brandColorsLength // This is to kick us out of the inner loop, since we've found a good match
      }
    }
  }

  updatePreviewPin () {
    let _this = this
    this.previewPinIsUpdating = true

    // This is manually debouncing cursor position updates
    window.setTimeout(function () { _this.setState({ previewPinIsUpdating: false }) }, 50)

    // Clear the previous mouse move call to findBrandColor and set a new one for the current cursor position
    window.clearTimeout(this.findBrandColorTimeout)
    this.findBrandColorTimeout = window.setTimeout(this.findBrandColor, 700)
  }

  // @todo - fix coords -RS
  wrapperMouseMove (e) {
    if (!this.state.previewPinIsUpdating) {
      const cursorX = e.clientX - this.wrapperOffset.left // can also try screenX or pageX
      const cursorY = e.clientY - this.wrapperOffset.top
      // const mappedCanvasIndex = (cursorY * this.state.canvasWidth + cursorX) * 4
      // Using a bitwise or to trunc here since it is much faster than something like Math.trunc()
      const mappedCanvasIndex = (((e.clientY - this.canvasOffset.top) * this.state.canvasWidth) + (e.clientX - this.canvasOffset.left | 0)) * 4
      // eslint-disable-next-line react/prop-types
      const imageData = this.props.data.imageData.data

      this.setState({
        currentPixelRGB: [imageData[mappedCanvasIndex], imageData[mappedCanvasIndex + 1], imageData[mappedCanvasIndex + 2]],
        currentPixelRGBstring: 'rgb(' + imageData[mappedCanvasIndex] + ',' + imageData[mappedCanvasIndex + 1] + ',' + imageData[mappedCanvasIndex + 2] + ')',
        mappedCanvasIndex: mappedCanvasIndex,
        previewPinIsUpdating: true,
        previewPinIsActive: false,
        previewPinX: cursorX,
        previewPinY: cursorY }, () => {
        (() => { this.updatePreviewPin() })()
      })
    }
  }

  componentDidMount () {
    this.initCanvas()
  }

  componentDidUpdate (prevProps, prevState) {
    // eslint-disable-next-line react/prop-types
    if (prevProps.width !== this.props.width) {
      // eslint-disable-next-line react/prop-types
      this.handleResize(prevProps.width, prevProps.height)
    }
  }

  render () {
    // eslint-disable-next-line react/prop-types
    const { isPortrait, img } = this.props.data

    const { previewPinIsActive, previewPinX, previewPinY, previewColorName, previewColorNumber, currentPixelRGBstring } = this.state
    const transformValue = `translate(${previewPinX}px, ${previewPinY}px)`

    return (
      <Fragment>
        <img ref={this.imageRef} className={hiddenImageClass} src={img} onLoad={this.handleImageLoaded} alt='hidden' />
        <div onMouseMove={this.wrapperMouseMove} className={wrapperClass} ref={this.CFIWrapper}>
          {/* eslint-disable-next-line react/prop-types */}
          <canvas className={isPortrait ? portraitOrientation : canvasClass} ref={this.CFICanvas} width={this.props.width} height={this.props.height} />
          <ColorsFromImagePin key='1'
            isActiveFlag={previewPinIsActive}
            pinType='preview'
            previewColorName={previewColorName}
            previewColorNumber={previewColorNumber}
            RGBstring={currentPixelRGBstring}
            transformValue={transformValue}
            onClickMethod={this.addPin} />
        </div>
      </Fragment>
    )
  }
}

export default ColorsFromImage

/*
-------------------------------------------------------------------------------------------
------------------ ### TO DO ### ------------------
-------------------------------------------------------------------------------------------
Do three steps of increasing fidelity checking in findBrandColor

Memoize result of findBrandColor

Confirm that I'm manually debouncing mousemove
*/
