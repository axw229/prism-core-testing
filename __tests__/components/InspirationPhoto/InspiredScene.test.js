import React from 'react'
import { shallow } from 'enzyme'
import InspiredScene from 'src/components/InspirationPhotos/InspiredScene'
import ColorsFromImage from 'src/components/InspirationPhotos/ColorsFromImage'
let defaultProps = {
  data: {},
  isActivedPage: true,
  isRenderingPage: true
}

const createInspiredScene = (props) => {
  let newProps = Object.assign({}, defaultProps, props)
  return shallow(<InspiredScene {...newProps} />)
}

const wrapper = createInspiredScene()

describe('compoments rendring test', () => {
  it('should rendering ColorsFromImage Component', () => {
    expect(wrapper.find(ColorsFromImage).exists()).toBe(true)
  })
})
