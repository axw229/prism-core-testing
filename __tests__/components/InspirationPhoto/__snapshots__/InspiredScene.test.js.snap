// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot match testing snapshot test 1`] = `
<div
  className="scene__image__wrapper"
>
  <ColorsFromImage
    data={Object {}}
    isActivedPage={true}
  />
</div>
`;
